[webkitbugspy] Set keywords on issues

JonWBedard opened the pull request:
#### 9c049cef55a9dd9b034e7fbc5213ff3117b59fe0
<pre>
[webkitbugspy] Set keywords on issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=260184">https://bugs.webkit.org/show_bug.cgi?id=260184</a>
<a href="https://rdar.apple.com/113880142">rdar://113880142</a>

Reviewed by Dewei Zhu.

Tools and services should be able to add and remove keywords from
radars and bugzillas.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.set): Add ability to set keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.set_keywords): Set keywords in tracker implementation.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue): Mock adding and setting keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.keywords): Pull from pre-constructed Keyword objects.
(RadarModel.remove_keyword): Remove a keyword from the issue.
(RadarModel.add_keyword): Add a keyword to the issue.
(RadarClient.keywords_for_name): List all keywords starting with a given string.
(Radar.__init__): Construct set of available keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.__init__): Keep track of previously queries keywords.
(Tracker.set): Add and remove keywords from a radar.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: <a href="https://commits.webkit.org/270193@main">https://commits.webkit.org/270193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a05ae006d2f9c6cd816ac5b61dbddda09baf0fd8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26941 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25091 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/5050 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/804 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25067 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/5050 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/21429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27524 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/24987 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/5050 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/22360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/5050 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/26338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2055 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/3344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2505 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3164 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
```diff
@@ -159,6 +159,13 @@ def _issue(self, url, id, credentials=None, data=None):
+                if keywords.get('set'):
+                    issue['keywords'] = keywords.get('set')
+                elif keywords.get('add'):
+                    issue['keywords'] = issue.get('keywords', []) + keywords.get('add')
```
It looks like `set` has higher priority than `add`. I guess we only allow one key in `keywords` dictionary.
Maybe we should `assert len(keywords) == 1` and then logging / throwing error for unsupported keys.
In this way, we can check against malformed `keywords`.

JonWBedard replied:
Want to note that this is actually a mock. That being said, an `assert` here would be nice.

dewei-zhu commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
```diff
@@ -387,6 +396,12 @@ def clone_radar(self, problem_id, reason_text, component=None):
             reproducible='Always',
         ))
 
+    def keywords_for_name(self, keyword_name):
```
It looks like this is actually `keywords_for_prefix` or `keywords_with_prefix`?

JonWBedard replied:
It probably should be `keywords_for_prefix`, but this is actually a mock of a `radarclient` function, which behaves as I've implemented in this mock.

dewei-zhu commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
```diff
@@ -358,6 +361,26 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
+                        self._keywords[candidate.name] = candidate
+                if keyword in self._keywords:
+                    continue
+                self._invalid_keywords.add(keyword)
+                raise ValueError("'{}' is not a valid keyword".format(keyword))
```
It looks like we raise the exception the first time we add keyword to `_invalid_keywords`, that means we are always checking against empty `_invalid_keywords` on line 366. If that's the case, do we still need this variable?

JonWBedard replied:
Checking for keywords is actually a very expensive network call. Presumably, our caller will be handling this exception, and in that case, we can save an expensive network call if our caller re-attempts the same invalid keyword in a single program. That's actually a relatively common event in practice, where a program is likely to have keywords constructed from variables and then re-used on multiple issues.

dewei-zhu reviewed (approved):
r=me with comment.

JonWBedard committed:
[webkitbugspy] Set keywords on issues
https://bugs.webkit.org/show_bug.cgi?id=260184
rdar://113880142

Reviewed by Dewei Zhu.

Tools and services should be able to add and remove keywords from
radars and bugzillas.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.set): Add ability to set keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.set_keywords): Set keywords in tracker implementation.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue): Mock adding and setting keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.keywords): Pull from pre-constructed Keyword objects.
(RadarModel.remove_keyword): Remove a keyword from the issue.
(RadarModel.add_keyword): Add a keyword to the issue.
(RadarClient.keywords_for_name): List all keywords starting with a given string.
(Radar.__init__): Construct set of available keywords.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.__init__): Keep track of previously queries keywords.
(Tracker.set): Add and remove keywords from a radar.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: https://commits.webkit.org/270193@main
